src/undo_log.rs

```diff
@@ -0,0 +1,225 @@
+//! Module which contains the snapshot/rollback functionality of the `ena` data structures.
+//!
+//! For most usecases this is just an internal implementation detail. However if many `ena`
+//! data structures are used snapshotted simultaneously it is possible to use
+//! `UnificationTableStorage`/`SnapshotVecStorage` instead and use a custom `UndoLogs<T>`
+//! type capable of recording the actions of all used data structures.
+//!
+//! Since the `*Storage` variants do not have an undo log `with_log` must be called with the
+//! unified log before any mutating actions.
+
+/// A trait which allows actions (`T`) to be pushed which which allows the action to be undone at a
+/// later time if needed
+pub trait UndoLogs<T> {
+    fn in_snapshot(&self) -> bool {
+        self.num_open_snapshots() > 0
+    }
+    fn num_open_snapshots(&self) -> usize {
+        0
+    }
+    fn push(&mut self, undo: T);
+    fn clear(&mut self);
+    fn extend<I>(&mut self, undos: I)
+    where
+        Self: Sized,
+        I: IntoIterator<Item = T>,
+    {
+        undos.into_iter().for_each(|undo| self.push(undo));
+    }
+}
+
+impl<'a, T, U> UndoLogs<T> for &'a mut U
+where
+    U: UndoLogs<T>,
+{
+    fn in_snapshot(&self) -> bool {
+        (**self).in_snapshot()
+    }
+    fn num_open_snapshots(&self) -> usize {
+        (**self).num_open_snapshots()
+    }
+    fn push(&mut self, undo: T) {
+        (**self).push(undo)
+    }
+    fn clear(&mut self) {
+        (**self).clear();
+    }
+    fn extend<I>(&mut self, undos: I)
+    where
+        Self: Sized,
+        I: IntoIterator<Item = T>,
+    {
+        (**self).extend(undos)
+    }
+}
+
+/// A trait which allows snapshots to be done at specific points. Each snapshot can then be used to
+/// rollback any changes to an underlying data structures if they were not desirable.
+///
+/// Each snapshot must be consumed linearly with either `rollback_to` or `commit`.
+pub trait Snapshots<T>: UndoLogs<T> {
+    type Snapshot;
+    fn has_changes(&self, snapshot: &Self::Snapshot) -> bool {
+        !self.actions_since_snapshot(snapshot).is_empty()
+    }
+    fn actions_since_snapshot(&self, snapshot: &Self::Snapshot) -> &[T];
+
+    fn start_snapshot(&mut self) -> Self::Snapshot;
+    fn rollback_to<R>(&mut self, values: impl FnOnce() -> R, snapshot: Self::Snapshot)
+    where
+        R: Rollback<T>;
+
+    fn commit(&mut self, snapshot: Self::Snapshot);
+}
+
+impl<T, U> Snapshots<T> for &'_ mut U
+where
+    U: Snapshots<T>,
+{
+    type Snapshot = U::Snapshot;
+    fn has_changes(&self, snapshot: &Self::Snapshot) -> bool {
+        (**self).has_changes(snapshot)
```

Comment:
Nit: I prefer `U::has_changes(..)` for these sorts of things; clearer, less error-prone